Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTP1Connections.swift

```diff
@@ -0,0 +1,438 @@
+                    continue
+                }
+
+                if conn.eventLoop === preferredEL {
+                    switch eventLoopMatch {
+                    case .none:
+                        eventLoopMatch = (index, connReturn)
+                    case .some((_, let existingMatchReturn)) where connReturn > existingMatchReturn:
+                        eventLoopMatch = (index, connReturn)
+                    default:
+                        break
+                    }
+                } else {
+                    switch goodMatch {
+                    case .none:
+                        goodMatch = (index, connReturn)
+                    case .some((_, let existingMatchReturn)):
+                        // We don't require a specific eventLoop. For this reason we want to pick a
+                        // matching eventLoop that has been idle the shortest.
+                        if connReturn > existingMatchReturn {
+                            goodMatch = (index, connReturn)
+                        }
+                    }
+                }
+            }
+
+            if let (index, _) = eventLoopMatch {
+                return index
+            }
+
+            if let (index, _) = goodMatch {
+                return index
+            }
+
+            return nil
+        }
+
+        func findAvailableConnection(onRequired requiredEL: EventLoop) -> Int? {
+            var match: (Int, NIODeadline)?
+
+            // To find an appropriate connection we iterate all existing connections.
+            // While we do this we try to find the best fitting connection for our request.
+            //
+            // A match, runs on the same eventLoop and has been idle the shortest amount of time.
+            for (index, conn) in self.connections.enumerated() {
+                // 1. Ensure we are on the correct EL.
+                guard conn.eventLoop === requiredEL else {
+                    continue
+                }
+
+                // 2. Ensure the connection is available
+                guard let connReturn = conn.availableAndLastReturn else {
+                    continue
+                }
+
+                switch match {
+                case .none:
+                    match = (index, connReturn)
+                case .some((_, let existingMatchReturn)) where connReturn > existingMatchReturn:
+                    // the currently iterated eventLoop has been idle for a shorter amount than
+                    // the current best match.
+                    match = (index, connReturn)
+                default:
+                    // the currently iterated eventLoop has been idle for a longer amount than
+                    // the current best match. We continue the iteration.
+                    continue
+                }
+            }
+
+            if let (index, _) = match {
+                return index
+            }
+
+            return nil
```

Comment:
Very subjective in terms of neatness, but can't you just `return match?.0`?

Reply:
Yes, I could do this. However my brain is often very slow. And parsing the shorthand syntax and translating it into the above takes a bit of time. For this reason, I personally prefer to go the verbose way. Not a hill I want to die on.